src/object/obj_tx.c

```diff
@@ -1993,23 +2034,62 @@ dc_tx_restart(tse_task_t *task)
 {
 	daos_tx_restart_t	*args;
 	struct dc_tx		*tx;
+	uint32_t		 backoff = 0;
 	int			 rc = 0;
 
 	args = dc_task_get_args(task);
 	D_ASSERTF(args != NULL,
 		  "Task Argument OPC does not match DC OPC (restart)\n");
 
-	tx = dc_tx_hdl2ptr(args->th);
+	tx = tse_task_get_priv(task);
 	if (tx == NULL) {
-		rc = -DER_NO_HDL;
-	} else {
-		rc = dc_tx_restart_internal(tx);
+		/* Executing task for the first time. */
 
-		/* -1 for hdl2ptr */
+		tx = dc_tx_hdl2ptr(args->th);
+		if (tx == NULL) {
+			rc = -DER_NO_HDL;
+			goto out;
+		}
+
+		D_MUTEX_LOCK(&tx->tx_lock);
+
+		rc = dc_tx_restart_begin(tx, &backoff);
+		if (rc != 0)
+			goto out_tx_lock;
+
+		if (backoff == 0) {
+			dc_tx_restart_end(tx);
+		} else {
+			/*
+			 * Reinitialize task with a delay to implement the
+			 * backoff and call dc_tx_restart_end.
+			 */
+			rc = tse_task_reinit_with_delay(task, backoff);
+			if (rc != 0) {
+				/* Skip the backoff. */
+				backoff = 0;
+				dc_tx_restart_end(tx);
+				goto out_tx_lock;
+			}
+			D_MUTEX_UNLOCK(&tx->tx_lock);
+			/* Pass our tx reference to task. */
+			tse_task_set_priv(task, tx);
+			return 0;
+		}
+
+out_tx_lock:
+		D_MUTEX_UNLOCK(&tx->tx_lock);
 		dc_tx_decref(tx);
+	} else {
+		/* Re-executing task after the reinitialization. */
```

Comment:
Some confused, under which case the dc_tx_restart() will be re-executed? The caller call dc_tx_restart() repeatedly  or the TX needs to be restart again? But for both cases, the "D_ASSERTF(tx->tx_status == TX_RESTARTING, "%d\n", tx->tx_status);" check in dc_tx_restart_end() may be not true.

Reply:
"Re-executing task after the reinitialization" means task is executed again because of the tse_task_reinit_with_delay call above.

Reply:
This is why I think using a child task instead of reinitializing task is easier to understand.

Reply:
There _is_ a mistake in using tse_task_set_priv. I should have used tse_task_set_priv_internal instead.